pkg/vgmanager/vgmanager_controller.go

```diff
@@ -76,6 +82,37 @@ func (r *VGReconciler) reconcile(ctx context.Context, req ctrl.Request) (ctrl.Re
 		return ctrl.Result{}, err
 	}
 
+	// cleanup on current node when CR is deleted
+	if !lvmCluster.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(lvmCluster, vgFinalizer) {
+			// TODO: any possibility to make vgCleanup async?
+			err := vgCleanup(r, ctx)
+			if err != nil {
+				r.Log.Error(err, "failed to perform VG Cleanup on node %s", nodeName)
+				return ctrl.Result{}, err
+			}
+
+			// remove finalizer after cleanup
+			controllerutil.RemoveFinalizer(lvmCluster, vgFinalizer)
+			if err := r.Client.Update(ctx, lvmCluster); err != nil {
+				r.Log.Error(err, "failed to remove finalizer", "node", nodeName, "CR", lvmCluster.Name)
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// set node uid as finalizer on CR
+	if !controllerutil.ContainsFinalizer(lvmCluster, vgFinalizer) {
```

Comment:
Can't we use `NodeName` itself as the finalizer ?

Reply:
- IMHO, UID is the standard way for these type of tasks.
- We can customize the finalizer to have prefix `node-uid/<uid>` for easy ref